src/Common/Core/Impl/Shell/CoreShellExtensions.cs

```diff
@@ -6,12 +6,16 @@
 using System.Runtime.CompilerServices;
 using System.Threading;
 using System.Threading.Tasks;
+using System.Windows.Threading;
 using Microsoft.Common.Core.Threading;
 
 namespace Microsoft.Common.Core.Shell {
     public static class CoreShellExtensions {
-        public static MainThreadAwaitable SwitchToMainThreadAsync(this ICoreShell coreShell, CancellationToken cancellationToken = default(CancellationToken))
-            => ((IMainThread)coreShell).SwitchToAsync(cancellationToken);
+        public static MainThreadAwaitable SwitchToMainThreadAsync(this ICoreShell coreShell, CancellationToken cancellationToken = default(CancellationToken)) {
+            var disp = Dispatcher.FromThread(coreShell.MainThread);
+            Debug.Assert(!disp.HasShutdownStarted, "Unable to transition to UI thread: dispatcher has started shutdown.");
```

Comment:
This should be part of `IMainThread` implementation. And I think we should actually throw exception, not just do `Debug.Assert`